### ISSUE TITLE
Diffing algorithm is sometimes hard to parse

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#4682]
Many times diffing is super helpful but sometimes it is very hard to understand exactly what is wrong:

<img width="997" alt="screen shot 2016-05-24 at 22 57 01" src="https://cloud.githubusercontent.com/assets/9582/15519674/da18c3ce-2202-11e6-992c-8744f3ad4ab4.png">

The diff helps because it gives me an idea of *where* is wrong but it is hard to put together exactly *what*. I have been thinking about this and what if we print the diff alongside the original lhs and rhs instead of a separate row? Something like this:

![image](https://cloud.githubusercontent.com/assets/9582/15519893/ebe11632-2203-11e6-82b9-f101d5e103e2.png)

For maps, the lhs would show in red values that it is not on the lhs and the rhs in green values not in the lhs. Same keys different values follow the same rule. Or maybe for maps the diff: row is the best way to go still.

Thoughts?

To-do list taken from https://github.com/elixir-lang/elixir/issues/4682#issuecomment-221511114:

* [ ] 1. List diff needs improvement – make it less noisy and more content aware
* [x] 2. Handle whitespace diff in some way (thoughts?)
* [x] 3. Try separated diff

**Comment by antipax** [comment]
👍 . Also, whitespace diffs are not highlighted in any way, which has been annoying a couple of times.

**Comment by josevalim** [comment]
@lexmag maybe one solution is to have the format diff function return a diff list, e.g. `[eq: ..., less: ..., more: ..., [less: ...]]` which can be arbitrarily nested. Then we print the lhs with eq and removals and the rhs with eq and additions. Thoughts @lexmag? :)

**Comment by lexmag** [comment]
I think there are 3 problems mentioned in the issue:
1. List diff needs improvement – make it less noisy and more content aware
2. Handle whitespace diff in some way (thoughts?)
3. Try separated diff

I'm starting work on 1. and 3.
I think having separated diff will be really good improvement, almost every diff tool provide both, unified and separated, diffing but I don't remember when I've used unified diff last time in such tools.
It should also solve the issue with inner string diff after https://github.com/elixir-lang/elixir/commit/bdee464c0cdb5fafb1ba2fd3f0fc28f03ffffb94.

**Comment by antipax** [comment]
I think the only whitespace diff that isn't handled is spaces, so maybe we can display the diff for those as spaces with a green or red background?

**Comment by lpil** [comment]
Traditionally an interpunct character is used to signify whitespace.
https://en.wikipedia.org/wiki/Interpunct

I could have a go at implementing this or the coloured background solution if that's OK.

**Comment by josevalim** [comment]
@lexmag I think I found a bug in the new diff!

![image](https://cloud.githubusercontent.com/assets/9582/15629582/e969bc3a-251c-11e6-83c3-1335d2d86904.png)

The comma should be part of the diff most likely. :)

**Comment by lexmag** [comment]
@josevalim the fix is in master.

**Comment by alco** [comment]
Sometimes it just looks cute
<img width="665" alt="screenshot 2016-06-03 14 15 42" src="https://cloud.githubusercontent.com/assets/207748/15777134/c9b289a8-2995-11e6-877e-ba7f65ca9d8c.png">

**Comment by lexmag** [comment]
Alright, there are first results for content-aware lists difference:

* first pair of lines – current output
* second pair of lines – new output

_ignore keywords formatting_

<img width="368" alt="screen shot 2016-06-05 at 00 15 56" src="https://cloud.githubusercontent.com/assets/248290/15802468/afeb2592-2ab3-11e6-9df0-34ed5de25d61.png">
